Prepare CI for NB 22 and bump minimum JDK requirement to 17

mbien opened the pull request:
- bump JDK from 11 to 17 as new minimum requirement for NB 22
   - some tests have to stay on 11 which require the nashorn engine
   - also see #4904 for the remaining JDK 8 locked tests
   - the version bump does also allow more tests to run, some needed fixing e.g ConvertToRecordPatternTest and ConvertToNestedRecordPatternTest
   - this is according to the agreed [Minimum JDK build and run policy](https://cwiki.apache.org/confluence/display/NETBEANS/Minimum+JDK+build+and+run+policy)
 - bump action versions to fix deprecation warnings
   - actions/cache to v4
   - actions/setup-node to v4
 
note: the graal job is still using an JDK 11 equivalent since #6369, this PR does not change this. However, this should unblock future graal updates, see https://github.com/apache/netbeans/pull/7013#issuecomment-1916217918

mbien commented:
well. _some_ tests did in fact _not_ fail :)

mbien commented:
fascinating. Out of almost 3k `webcommon/javascript2.editor` tests, a single completion test is failing:
https://github.com/apache/netbeans/blob/805b153827a17d4412062e2397f1f0d094944e78/webcommon/javascript2.editor/test/unit/src/org/netbeans/modules/javascript2/editor/JsCodeCompletionGeneralTest.java#L67-L69

matthiasblaesing commented:
> fascinating. Out of almost 3k `webcommon/javascript2.editor` tests, a single completion test is failing:
> 
> https://github.com/apache/netbeans/blob/805b153827a17d4412062e2397f1f0d094944e78/webcommon/javascript2.editor/test/unit/src/org/netbeans/modules/javascript2/editor/JsCodeCompletionGeneralTest.java#L67-L69

I think that we are seeing a scanning effect here. Even for JS this is bogus:

https://github.com/apache/netbeans/blob/805b153827a17d4412062e2397f1f0d094944e78/webcommon/javascript2.editor/test/unit/data/testfiles/completion/issue215777.js#L1-L2

I assume that the indexer picks up the `Math.var` (End of line 1 + beginning of line 2) and reports that. The tested issue deals with the space between `=` and `Math`, so I suggest to split the two cases and this should clear up.

Patch (wrapped in ZIP because GH in its infitite wisdom rejects my patch file...)

[0001-Stabilize-JS-Test-testIssue215777.zip](https://github.com/apache/netbeans/files/14115839/0001-Stabilize-JS-Test-testIssue215777.zip)

I was able to reproduce the problem locally and this fixed it for me.

mbien commented:
thanks @matthiasblaesing, I added your commit - everything should be green now. The cause for this might be that the method order is slightly different.

mbien commented on harness/nbjunit/src/org/netbeans/junit/MethodOrder.java:
```diff
@@ -43,6 +43,9 @@ static void initialize() {
         String orderS = findOrder();
         if (!"natural".equals(orderS)) { // NOI18N
             try {
+                // TODO: ClassLoader fields can not be accessed via reflection on JDK 12+
+                // see jdk.internal.reflect.Reflection#fieldFilterMap
+                // this won't work
                 Field classesF = ClassLoader.class.getDeclaredField("classes"); // NOI18N
```
luckily nothing seems to be using test case ordering: https://github.com/search?q=repo%3Aapache%2Fnetbeans%20NbTestCase.order&type=code

so I haven't bothered to fix it.

mbien commented on java/java.hints.test/src/org/netbeans/modules/java/hints/test/api/HintTest.java:
```diff
@@ -381,12 +378,27 @@ private void ensureCompilable(FileObject file) throws IOException, AssertionErro
         }
     }
 
+    /**Sets a source level for all Java files used in this test.
+     *
+     * @param sourceLevel the source level to use while parsing Java files
+     * @return itself
+     */
+    public HintTest sourceLevel(int sourceLevel) {
+        assertTrue(sourceLevel >= 8);
+        return this.sourceLevel(Integer.toString(sourceLevel));
+    }
+
     /**Sets a source level for all Java files used in this test.
      *
      * @param sourceLevel the source level to use while parsing Java files
      * @return itself
      */
     public HintTest sourceLevel(String sourceLevel) {
+        try {
+            int valid = sourceLevel.startsWith("1.") ? Integer.parseInt(sourceLevel.substring(2)) : Integer.parseInt(sourceLevel);
+        } catch (NumberFormatException ex) {
+            throw new IllegalArgumentException(ex);
+        }
         this.sourceLevel = sourceLevel;
         return this;
     }
```
some tests used `SourceVersion.latest().name()` as source level which would be the full enum name. This would be quietly ignored and cause weird issues.

The expected format is now enforced and the tests were fixed which used it wrong.

mbien commented on java/java.hints/test/unit/src/org/netbeans/modules/java/hints/jdk/ConvertToRecordPatternTest.java:
```diff
@@ -92,7 +89,7 @@ public void testDuplicateVarName() throws Exception {
                         + "public class Test {\n"
                         + "    private int test(Object s) {\n"
-                        + "        if (s instanceof Person(String name, int s1) p) {\n"
+                        + "        if (s instanceof Person(String name, int s1)) {\n"
```
@lahodaj i think the tests were wrong? This doesn't seem to be a valid syntax when it also includes the field name.

Another side effect of this PR is that more tests are now active.

matthiasblaesing reviewed (approved):
Looks sane to me. Thank you.

mbien commented:
NB 22 will be released when JDK 22 is already out, I am wondering if we should:

a) change matrix to `[17. 21, 22ea]`
b) change matrix to `[17, 22ea]` since we can assume JDK 21 will do fine if it works on 22 (add 21 again once we are at 23ea)

I am slightly in favor of a) even though it wastes a bit more resources, since one release from now the delta between 21 and _current_ will grow, then we will have to run the a) variant anyway.

neilcsmith-net commented:
@mbien I'm also in favour of a) because that is the approach agreed in the minimum JDK support vote.

mbien commented:
- rebased on latest master and fixed conflicts from the just synced delivery branch
 - added 22-ea to the matrix

merging once green assuming nothing else shows up

mbien commented:
i think this will cause too many failures atm. The commit validation matrix would produce 5 runs, combined with the java debugger job at 3 runs we would have to play whack-a-mole a bit too much, I already restarted twice and it is failing again.

I am going to replace some 21 runs with 22-ea instead of adding 22-ea as extra run. Retry scripts are already in use so the option is spent.

mbien commented:
all green, merging
